### PR TITLE
Yevgeny/scrapy refactor

### PR DESCRIPTION
**Pull request by yevgenybulochnik**
Fixes coderxio/dailymed-api #7 #6 

## Explanation
This is a very large PR with a lot of adjustments. Going forward I would really like to have smaller more frequent PRs. This PR refactors the scrapy items + adds a scrapy pipeline for loading data into the db. Because of the scrapy adjustments I also had to refactor the models and views. This PR does not include active or inactive ingredients at this time (xpaths are commented out currently) 

@finish06 added some tests to the PR as well. I would like to make a proposal to use pytest vs djangos unittest. 

## Rationale
The scrapy pipeline is beneficial because it allows for better separation of concerns and makes data output processing more modular. The models and views needed to be adjusted to accommodate a Set > Spl > Product > Package setup, but these can all change. We will likely need many more serializers!   

## Tests
1. What testing did you do?
* Django unit tests run without issue
* Loaded first 50 spls without issue
* Browsable api shows correct views and responses (see attached spl output) 

<details>
<summary>testing logs</summary>

```
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
...
----------------------------------------------------------------------
Ran 3 tests in 0.003s

OK
{
    "id": "fe1df76d-4925-444d-8e30-9dd07f205fb7",
    "products": [
        {
            "id": 7,
            "packages": [
                {
                    "id": 12,
                    "code": "68788-7665-3"
                },
                {
                    "id": 13,
                    "code": "68788-7665-6"
                },
                {
                    "id": 14,
                    "code": "68788-7665-9"
                },
                {
                    "id": 15,
                    "code": "68788-7665-1"
                }
            ],
            "code": "68788-7665",
            "name": "Losartan Potassium"
        }
    ],
    "labeler": "Preferred Pharmaceuticals, Inc.",
    "set": "fe1df76d-4925-444d-8e30-9dd07f205fb7"
}
```
</details>

### CODE REVIEW

**Comment by yevgenybulochnik** [comment]
Forgot to mention, I did not run code formatting on this branch. I would suggest we make a separate PR with all the code formatting updates, and make sure we stay consistent after that.

**Comment by finish06** [comment]
I would agree the formatting changes should be a separate own branch.  I used this branch last night when writing the test cases.  I am comfortable with this merge.

File: api/dailymed/models.py [hunk 1]
```diff
@@ -5,20 +5,43 @@ class Set(models.Model):
...
+    def __str__(self):
+        return f"{self.id} -- {self.set} -- {self.labeler}"
 
-class Ndc(models.Model):
-    ndc = models.CharField(max_length=15)
+
+class Product(models.Model):
+    """Product level model"""
+    code = models.CharField(max_length=20)
+    name = models.CharField(max_length=50)
```
Comment (jrlegrand) [inline api/dailymed/models.py:29]: I think we are definitely going to want to make the product name field longer than 50 characters.  There's some lonnnng names for medication products.

File: api/dailymed/models.py [hunk 2]
```diff
@@ -5,20 +5,43 @@ class Set(models.Model):
...
     spl = models.ForeignKey(
         Spl,
         on_delete=models.PROTECT,
-        related_name="ndcs"
+        related_name='products',
+    )
+
+    def __str__(self):
+        return f"{self.code} -- {self.name} -- {self.spl}"
```
Comment (jrlegrand) [inline api/dailymed/models.py:37]: I am unfamiliar with this - is this related to "Black" formatting rules?

File: api/dailymed/serializers.py [hunk 3]
```diff
@@ -1,20 +1,34 @@
...
+from dailymed.models import Set, Spl, Product, Package
 
-class SetSerializer(serializers.HyperlinkedModelSerializer):
+
+class PackageSerializer(serializers.ModelSerializer):
     class Meta:
-        model = Set
-        fields = ['id', 'spls']
+        model = Package
+        exclude = ('product', )
```
Comment (jrlegrand) [inline api/dailymed/serializers.py:8]: What is the purpose of these "exclude"s?

File: api/dailymed/serializers.py [hunk 4]
```diff
@@ -1,20 +1,34 @@
...
+    products = ProductSerializer(many=True)
 
-class SplSerializer(serializers.HyperlinkedModelSerializer):
     class Meta:
         model = Spl
-        fields = ['id', 'set', 'ndcs']
-        extra_kwargs = {
-            'ndcs': {'lookup_field': 'ndc'}
-        }
+        fields = '__all__'
```
Comment (jrlegrand) [inline api/dailymed/serializers.py:24]: How is the "Set" relation represented in "Spl"?  Just the ID?  Or a Hyperlink?  Or other?

File: api/dailymed/urls.py [hunk 5]
```diff
@@ -6,9 +6,10 @@
 router = DefaultRouter()
 router.register(r'set', views.SetViewSet)
 router.register(r'spl', views.SplViewSet)
-router.register(r'ndc', views.NdcViewSet)
+router.register(r'product', views.ProductViewSet)
+router.register(r'ndc', views.PackageViewSet)
```
Comment (jrlegrand) [inline api/dailymed/urls.py:10]: Hmm...  having one route called product and another called ndc is a little odd.  But we can figure this out as we go along.  I also feel like renaming it to "package" feels weird too.

File: scraper/scraper/pipelines.py [hunk 6]
```diff
@@ -5,9 +5,19 @@
...
 
 # useful for handling different item types with a single interface
-from itemadapter import ItemAdapter
+from dailymed.models import Set
 
 
 class ScraperPipeline:
     def process_item(self, item, spider):
+        set_id = item.pop('set_id')
+        products_data = item.pop("products")
```
Comment (jrlegrand) [inline scraper/scraper/pipelines.py:14]: Double quotes (sorry I'm anal).

File: scraper/scraper/items.py [hunk 7]
```diff
@@ -3,16 +3,39 @@
...
-    ndc = scrapy.Field()
+class ProductItem(Item):
+    code = Field(
+        output_processor = TakeFirst()
+    )
+    name = Field(
+        output_processor = TakeFirst()
+    )
+    active_ing = Field()
+    inactive_ing = Field()
```
Comment (jrlegrand) [inline scraper/scraper/items.py:34]: I would prefer to spell out "ingredient" - DailyMed does in their API.  But no big deal.

**Review by jrlegrand — APPROVED** [review]
See random comments.  Nothing major.  Looks good!

File: api/dailymed/models.py (hunk 2 above)
Comment (finish06) [reply api/dailymed/models.py:37]: This is necessary for the test cases.  The test cases evaluate the data in the database using the builtin __str__ function.  Without this function, the test cases will not work.

File: api/dailymed/models.py (hunk 1 above)
Comment (yevgenybulochnik) [reply api/dailymed/models.py:29]: Only downside I could see, if size/performance is not impacted, is from a validation perspective.

File: api/dailymed/serializers.py (hunk 3 above)
Comment (yevgenybulochnik) [reply api/dailymed/serializers.py:8]: This exclude is really just to make the ProductViewset look a bit better, otherwise in the product view json response the product pk is shown in the product level data and in the ndc level data.

File: api/dailymed/models.py (hunk 1 above)
Comment (finish06) [reply api/dailymed/models.py:29]: I think we should use code for validation of input data, not DB errors.  Maybe go with 255, but use code to validate it is '50' or such.  This would provide enhanced expansion opportunities while also alert use to changes in the SPLs.

File: api/dailymed/urls.py (hunk 5 above)
Comment (yevgenybulochnik) [reply api/dailymed/urls.py:10]: Maybe we should make an issue about what end points we actually want to start off with. I feel like some of these views can be consolidated. I was initially going to use "package" but agree it sounds kinda weird. Maybe we just consolidate to spl and ndc?

File: scraper/scraper/items.py (hunk 7 above)
Comment (yevgenybulochnik) [reply scraper/scraper/items.py:34]: Ill adjust this, agree good to be more explicit.

File: scraper/scraper/pipelines.py (hunk 6 above)
Comment (yevgenybulochnik) [reply scraper/scraper/pipelines.py:14]: Will adjust this. So this is where a code formatter like [Black](https://github.com/psf/black) might help. One would run black on the code and it would auto-format everything. So in this case everything would get formatted to double quotes.

File: api/dailymed/serializers.py (hunk 4 above)
Comment (jrlegrand) [reply api/dailymed/serializers.py:24]: I think id is all we would need for now and the foreseeable future.

File: api/dailymed/urls.py (hunk 5 above)
Comment (jrlegrand) [reply api/dailymed/urls.py:10]: I'd have to think about that a little more.  I feel like I might want to use the API for a product NDC.

File: scraper/scraper/pipelines.py (hunk 6 above)
Comment (jrlegrand) [reply scraper/scraper/pipelines.py]: Bummer.  I'm more of a single quotes guy.  But then again I'm a tab guy as discussed so maybe I'm wrong about this as well.

**Review by finish06 — APPROVED** [review]
All final changes look good.
